### PR TITLE
Add first phase of FxA ETL for calculating MAU

### PR DESCRIPTION
**Pull request by jklukas**
There will be follow-ups in the next few days to include daily imports from the FxA team's BQ projects.

### CODE REVIEW

File: sql/firefox_accounts_exact_mau28_raw_v1.sql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  -- the last country observed for a user is considered for determining whether
+  -- they contribute to tier 1 MAU, but we also include an "inclusive" tier 1
+  -- mau calculation that we have previously been using for KPI calculations on
+  -- FxA data; a user is included in this calculation if it appeared in a tier 1
+  -- country during any of the 28 days in the MAU window.
+  COUNTIF(_inactive_days_tier1 < 28) AS mau_tier1_inclusive,
+  -- We hash user_ids into 20 buckets to aid in computing
+  -- confidence intervals for mau/wau/dau sums; the particular hash
+  -- function and number of buckets is subject to change in the future.
+  MOD(ABS(FARM_FINGERPRINT(user_id)), 20) AS id_bucket,
```
Comment (mreid-moz) [inline sql/firefox_accounts_exact_mau28_raw_v1.sql]: drive-by: should we shift to non-negative range rather than taking `ABS(...)` here?

Comment (jklukas) [reply sql/firefox_accounts_exact_mau28_raw_v1.sql]: I think it would be complex to do so. `FARM_FINGERPRINT` returns a signed INT64 and BQ does not have any larger or unsigned integer variants, so we don't have a type to shift into.

Is your concern that ABS is introducing a bias between buckets?

Comment (mreid-moz) [reply sql/firefox_accounts_exact_mau28_raw_v1.sql]: Yep, that's my concern. It may not matter, and I agree that working around it seems overly complex here.

**Comment by jklukas** [comment]
cc @zerokarmaleft

File: sql/firefox_accounts_exact_mau28_raw_v1.sql (hunk 1 above)
Comment (relud) [reply sql/firefox_accounts_exact_mau28_raw_v1.sql]: this would shift mod to positive, and guarantee we don't get a int overflow if we hit `-0x8000000000000000`:
```suggestion
  MOD(MOD(FARM_FINGERPRINT(user_id), 20) + 20, 20) AS id_bucket,
```

File: sql/fxa_users_daily_v1.sql [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    -- this table contains events exported from Amplitude and ends on 2019-03-17;
+    -- we only consider this table for now, but we will soon be unioning this with
+    -- data from a separate BigQuery project that does daily imports of FxA logs.
+    static.fxa_amplitude_export_event_date
+  WHERE
+    user_id IS NOT NULL
+    AND event_type NOT IN (
+      'fxa_email - bounced', 'fxa_email - click', 'fxa_email - sent',
+      'fxa_reg - password_blocked','fxa_reg - password_common', 'fxa_reg - password_enrolled',
+      'fxa_reg - password_missing','fxa_sms  - sent', 'mktg - email_click',
```
Comment (zerokarmaleft) [inline sql/fxa_users_daily_v1.sql]: Is the double space in `fxa_sms  - sent` a typo?

File: sql/fxa_users_last_seen_v1.init.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  submission_date AS date_last_seen,
+  IF(country IN ('United States', 'France', 'Germany', 'United Kingdom', 'Canada'),
+    submission_date,
+    NULL) AS date_last_seen_in_tier1_country,
+  * EXCEPT (submission_date, generated_time)
+FROM
+  fxa_users_daily_v1
+WHERE
+  -- 2017-10-01 is the first date in the Amplitude FxA project.
+  submission_date = DATE '2017-10-01'
```
Comment (relud) [inline sql/fxa_users_last_seen_v1.init.sql]: I think we should do this in the same format as `clients_last_seen_v1.init.sql`, so that it can be used to generate any date without prior days, which is useful during backfill.

```sql
WITH
  input AS (
  SELECT
    ROW_NUMBER() OVER w1_unframed AS _n,
    FIRST_VALUE(IF(country IN ('United States',
          'France',
          'Germany',
          'United Kingdom',
          'Canada'),
        submission_date,
        NULL) IGNORE NULLS) OVER w1 AS date_last_seen_in_tier1_country,
    *
  FROM
    analysis.fxa_users_daily_v1
  WHERE
    submission_date <= @submission_date
    AND submission_date > DATE_SUB(@submission_date, INTERVAL 28 DAY)
  WINDOW
    w1 AS (
    PARTITION BY
      user_id
    ORDER BY
      submission_date DESC ROWS BETWEEN UNBOUNDED PRECEDING
      AND UNBOUNDED FOLLOWING),
    -- We must provide a modified window for ROW_NUMBER which cannot accept a frame clause.
    w1_unframed AS (
    PARTITION BY
      user_id
    ORDER BY
      submission_date DESC) )
SELECT
  @submission_date AS submission_date,
  CURRENT_DATETIME() AS generated_time,
  submission_date AS last_seen_date,
  * EXCEPT (_n,
    submission_date,
    generated_time)
FROM
  input
WHERE
  _n = 1
```

File: sql/fxa_users_last_seen_v1.sql [hunk 4]
```diff
@@ -0,0 +1,47 @@
+WITH
+  current_sample AS (
+  SELECT
+    -- Include dummy dates for date_last_* fields to make schema match with `previous`.
+    DATE '2000-01-01' AS date_last_seen,
+    DATE '2000-01-01' AS date_last_seen_in_tier1_country,
```
Comment (relud) [inline sql/fxa_users_last_seen_v1.sql]: I think we should use `DATE(NULL)` here if it is a dummy column

File: sql/fxa_users_last_seen_v1.sql [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  WHERE
+    submission_date = DATE_SUB(@submission_date, INTERVAL 1 DAY)
+    AND date_last_seen > DATE_SUB(@submission_date, INTERVAL 28 DAY) )
+SELECT
+  @submission_date AS submission_date,
+  CURRENT_DATETIME() AS generated_time,
+  -- Record the last day on which we recieved any core ping at all from this user.
+  IF(current_sample.user_id IS NOT NULL,
+    @submission_date,
+    previous.date_last_seen) AS date_last_seen,
```
Comment (relud) [inline sql/fxa_users_last_seen_v1.sql]: I feel like we should use `COALESCE(current_sample.date_last_seen, previous.date_last_seen) AS date_last_seen` and do `@submission_date AS date_last_seen` on line 5

File: sql/fxa_users_last_seen_v1.sql [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  -- Record the last day on which the user was in a "Tier 1" country;
+  -- this allows a variant of country-segmented MAU where we can still count
+  -- a user that appeared in one of the target countries in the previous
+  -- 28 days even if the most recent "country" value is not in this set.
+  IF(current_sample.user_id IS NOT NULL
+  AND current_sample.country IN ('United States', 'France', 'Germany', 'United Kingdom', 'Canada'),
+    @submission_date,
+    IF(previous.date_last_seen_in_tier1_country > DATE_SUB(@submission_date, INTERVAL 28 DAY),
+      previous.date_last_seen_in_tier1_country,
+      NULL)) AS date_last_seen_in_tier1_country,
```
Comment (relud) [inline sql/fxa_users_last_seen_v1.sql]: I think we should use `COALESCE` here too

File: sql/firefox_accounts_exact_mau28_raw_v1.sql [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  CURRENT_DATETIME() AS generated_time,
+  COUNTIF(_inactive_days < 28) AS mau,
+  COUNTIF(_inactive_days < 7) AS wau,
+  COUNTIF(_inactive_days < 1) AS dau,
+  -- We are generally using an "exclusive dimensions" methodology where only
+  -- the last country observed for a user is considered for determining whether
+  -- they contribute to tier 1 MAU, but we also include an "inclusive" tier 1
+  -- mau calculation that we have previously been using for KPI calculations on
+  -- FxA data; a user is included in this calculation if it appeared in a tier 1
+  -- country during any of the 28 days in the MAU window.
```
Comment (relud) [inline sql/firefox_accounts_exact_mau28_raw_v1.sql]: note: this isn't quite accurate, due to `udf_mode_last`, but explaining it more precisely I think would make it the explanation too complicated. still, if you can think of a simple way to say `appeared in a tier_1 country more than any other country for at least one of the 28 days in the MAU window.` or something I think that would be better.

File: sql/firefox_accounts_exact_mau28_v1.sql [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  SUM(IF(country IN ('US', 'FR', 'DE', 'UK', 'CA'), mau, 0)) AS tier1_mau,
+  SUM(IF(country IN ('US', 'FR', 'DE', 'UK', 'CA'), wau, 0)) AS tier1_wau,
+  SUM(IF(country IN ('US', 'FR', 'DE', 'UK', 'CA'), dau, 0)) AS tier1_dau
+FROM
+  `moz-fx-data-derived-datasets.analysis.firefox_accounts_exact_mau28_by_dimensions_v1`
+GROUP BY
+  submission_date
+
+-- This is a "live view" and can be updated via bq:
+-- bq update --project moz-fx-data-derived-datasets --use_legacy_sql=false --view "$(cat sql/firefox_accounts_exact_mau28_v1.sql)" analysis.firefox_accounts_exact_mau28_v1
```
Comment (relud) [inline sql/firefox_accounts_exact_mau28_v1.sql]: I think people should prefer using stdin for these commands, just because command line arguments have lower length limits than stdin.
```suggestion
-- bq update --project moz-fx-data-derived-datasets --use_legacy_sql=false --view analysis.firefox_accounts_exact_mau28_v1 < sql/firefox_accounts_exact_mau28_v1.sql
```

File: sql/firefox_accounts_exact_mau28_raw_v1.sql (hunk 1 above)
Comment (jklukas) [reply sql/firefox_accounts_exact_mau28_raw_v1.sql]: I'm going to stick with the simpler form for now. I've broken out the hash function discussion to https://github.com/mozilla/bigquery-etl/issues/36 so it can be evaluated as a separate project in the future.

File: sql/fxa_users_last_seen_v1.sql (hunk 4 above)
Comment (jklukas) [reply sql/fxa_users_last_seen_v1.sql]: Makes sense. I didn't know DATE(NULL) was possible.

File: sql/fxa_users_last_seen_v1.sql (hunk 5 above)
Comment (jklukas) [reply sql/fxa_users_last_seen_v1.sql]: I'm going to think some more about this. It might make sense to wrap up some logic in a SQL UDF to make sure we're consistent between the different date_last_seen columns.

File: sql/fxa_users_daily_v1.sql (hunk 2 above)
Comment (jklukas) [reply sql/fxa_users_daily_v1.sql]: Indeed this is a bug.

File: sql/firefox_accounts_exact_mau28_v1.sql (hunk 8 above)
Comment (relud) [reply sql/firefox_accounts_exact_mau28_v1.sql]: you are totally right, I misread the help page and didn't test thoroughly enough >_<

File: sql/firefox_accounts_exact_mau28_raw_v1.sql [hunk 9]
```diff
@@ -18,8 +18,9 @@ SELECT
   -- the last country observed for a user is considered for determining whether
   -- they contribute to tier 1 MAU, but we also include an "inclusive" tier 1
   -- mau calculation that we have previously been using for KPI calculations on
-  -- FxA data; a user is included in this calculation if it appeared in a tier 1
-  -- country during any of the 28 days in the MAU window.
+  -- FxA data; we assign a single country per user per day and include a user
+  -- in this calculation if they were assigned a tier 1 country in any of the
+  -- 28 days of the MAU window.
```
Comment (relud) [inline sql/firefox_accounts_exact_mau28_raw_v1.sql]: ❤️ this is excellent

File: sql/fxa_users_last_seen_v1.sql (hunk 5 above)
Comment (jklukas) [reply sql/fxa_users_last_seen_v1.sql]: As I played with this, I better understood how your COALESCE proposal fits together and I've gone with that rather than a UDF.

File: sql/fxa_users_last_seen_v1.init.sql (hunk 3 above)
Comment (jklukas) [reply sql/fxa_users_last_seen_v1.init.sql]: We discussed face-to-face and I'm going to merge this init.sql as-is with the understanding that @relud is working on a new approach to schema definition and metadata for this repo which will make this init.sql irrelevant.

File: sql/firefox_accounts_exact_mau28_by_dimensions_v1.sql [hunk 10]
```diff
@@ -0,0 +1,18 @@
+SELECT
```
Comment (relud) [inline sql/firefox_accounts_exact_mau28_by_dimensions_v1.sql]: Instead of a comment at the bottom, we could do this as DDL
```suggestion
CREATE OR REPLACE VIEW
  `firefox_accounts_exact_mau28_by_dimensions_v1` AS
SELECT
```

Comment (jklukas) [reply sql/firefox_accounts_exact_mau28_by_dimensions_v1.sql]: Love it.
